homeassistant/components/tuya/cover.py

```diff
@@ -0,0 +1,137 @@
+"""Support for Tuya Cover."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+from homeassistant.components.cover import (
+    DEVICE_CLASS_CURTAIN,
+    DOMAIN as DEVICE_DOMAIN,
+    SUPPORT_CLOSE,
+    SUPPORT_OPEN,
+    SUPPORT_SET_POSITION,
+    SUPPORT_STOP,
+    CoverEntity,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity import Entity
+
+from .base import TuyaHaEntity
+from .const import (
+    DOMAIN,
+    TUYA_DEVICE_MANAGER,
+    TUYA_DISCOVERY_NEW,
+    TUYA_HA_DEVICES,
+    TUYA_HA_TUYA_MAP,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+TUYA_SUPPORT_TYPE = {"cl", "clkg"}  # Curtain  # Curtain Switch
+
+# Curtain
+# https://developer.tuya.com/en/docs/iot/f?id=K9gf46o5mtfyc
+DPCODE_CONTROL = "control"
+DPCODE_PERCENT_CONTROL = "percent_control"
+DPCODE_PERCENT_STATE = "percent_state"
+DPCODE_SITUATION_SET = "situation_set"
+
+ATTR_POSITION = "position"
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities
+) -> None:
+    """Set up tuya cover dynamically through tuya discovery."""
+    hass.data[DOMAIN][entry.entry_id][TUYA_HA_TUYA_MAP][
+        DEVICE_DOMAIN
+    ] = TUYA_SUPPORT_TYPE
+
+    async def async_discover_device(dev_ids):
+        """Discover and add a discovered tuya cover."""
+        _LOGGER.debug("cover add-> %s", dev_ids)
+        if not dev_ids:
+            return
+        entities = await hass.async_add_executor_job(
+            _setup_entities, hass, entry, dev_ids
+        )
+        async_add_entities(entities)
+
+    async_dispatcher_connect(
+        hass, TUYA_DISCOVERY_NEW.format(DEVICE_DOMAIN), async_discover_device
+    )
+
+    device_manager = hass.data[DOMAIN][entry.entry_id][TUYA_DEVICE_MANAGER]
+    device_ids = []
+    for (device_id, device) in device_manager.device_map.items():
+        if device.category in TUYA_SUPPORT_TYPE:
+            device_ids.append(device_id)
+    await async_discover_device(device_ids)
+
+
+def _setup_entities(
+    hass: HomeAssistant, entry: ConfigEntry, device_ids: list
+) -> list[Entity]:
+    """Set up Tuya Cover."""
+    device_manager = hass.data[DOMAIN][entry.entry_id][TUYA_DEVICE_MANAGER]
+    entities: list[Entity] = []
+    for device_id in device_ids:
+        device = device_manager.device_map[device_id]
+        if device is None:
+            continue
+        entities.append(TuyaHaCover(device, device_manager))
+        hass.data[DOMAIN][entry.entry_id][TUYA_HA_DEVICES].add(device_id)
+    return entities
+
+
+class TuyaHaCover(TuyaHaEntity, CoverEntity):
+    """Tuya Switch Device."""
+
+    _attr_device_class = DEVICE_CLASS_CURTAIN
+
+    @property
+    def is_closed(self) -> bool | None:
+        """Return is cover is closed."""
+        return None
```

Comment:
Why does this just return `None`?

Reply:
is_closed is not supported for Tuya curtain device

Reply:
If we know the current position we should be able to know if the cover is closed or not. Why is that not the case?

Reply:
We can't know position 100 is closed or opened now.

Reply:
Please explain more. The Home Assistant definition says that position 0 is closed. The platform must implement the cover entity to follow this definition.

https://developers.home-assistant.io/docs/core/entity/cover#platform-properties-to-be-implemented-by-deriving-platform-classes

Reply:
Because part of the position of the tuya cover is 100, which means closed, and part of position0 means closed, and it is currently impossible to judge according to the category. There is no particularly good way to determine this value in the home assistant.

Reply:
According to https://developer.tuya.com/en/docs/iot/s?id=K9gf48qy7wkre#title-1-Support%20reporting%20instructions%20to%20the%20cloud%20only you can do this:
```suggestion
        return self.tuya_device.status.get(DPCODE_SITUATION_SET) == "fully_closed"
```

Reply:
Oh sorry, just read the other comment. It's only to indicate if 0 or 100 is closed or not.

So something like this could work?
```suggestion
        return self.current_cover_position == 0
```